### PR TITLE
Support GHC-9.4

### PR DESCRIPTION
**Pull request by TeofilC**
This is an early attempt to get this working with GHC-9.4.1.

At the moment we need to use some patched versions of libraries taken from https://github.com/yesodweb/yesod/pull/1769

There's also an issue with changes to the constraint solver and TypeErrors (upstream issue: https://gitlab.haskell.org/ghc/ghc/-/issues/21677). I've worked around it by applying some TypeApplications.

This is mostly just here to save work for others who want to try using 9.4

### CODE REVIEW

**Comment by TeofilC** [comment]
Now that 9.4 has been released I'll try to tidy this up in the next couple of days

**Comment by TeofilC** [comment]
Currently everything builds except `servant-auth-swagger` and `servant-auth-docs` as they require `doctest` to build and afaik that's not been updated to work with 9.4. All the tests that can be compiled pass though.

I feel like the changes to `cabal.project` should be dropped before merging, but I've left them there for now to ease testing.

File: cabal.project [hunk 1]
```diff
@@ -51,3 +51,41 @@ packages:
...
 optimization: False
 -- reorder-goals: True
+
+allow-newer: all
+
+-- https://github.com/GaloisInc/cereal/pull/103
+source-repository-package
+    type: git
+    location: https://github.com/parsonsmatt/cereal
+    tag: b4acf0b778a1d5da305f4de013ce7f32b53d8282
```
Comment (ysangkok) [inline cabal.project:61]: this can be removed, see https://github.com/GaloisInc/cereal/commits/master

File: cabal.project [hunk 2]
```diff
@@ -51,3 +51,41 @@ packages:
...
+    subdir: foundation
+
+-- https://github.com/vincenthz/hs-memory/pull/93
+source-repository-package
+    type: git
+    location: https://github.com/parsonsmatt/hs-memory
+    tag: 296b79424854eae293f6ba09b5308a0bf4dfd6d5
+
+
+-- https://github.com/snap-framework/io-streams/pull/83
```
Comment (ysangkok) [inline cabal.project:78]: Link is wrong, should be https://github.com/snapframework/io-streams/pull/83

**Comment by alpmestan** [comment]
@TeofilC Thanks a lot for this. I just allowed CI to run for this branch. I'll try to keep an eye on this PR, let me know when you consider this ready (i.e the doctest business is sorted out etc).

Cheers!

File: cabal.project (hunk 1 above)
Comment (TeofilC) [reply cabal.project]: Nice thanks

File: cabal.project (hunk 2 above)
Comment (TeofilC) [reply cabal.project]: Cheers that's been merged now so I've removed the line

**Comment by TeofilC** [comment]
Still waiting on some things but I've updated the overrides and listed the allow-newers

**Comment by benz0li** [comment]
Cross reference: https://github.com/haskell-servant/servant/issues/1608

File: cabal.project [hunk 3]
```diff
@@ -48,6 +48,36 @@ packages:
...
-tests: True
+tests: False
 optimization: False
 -- reorder-goals: True
+
+-- https://github.com/vincenthz/hs-memory/pull/93
+source-repository-package
+    type: git
+    location: https://github.com/parsonsmatt/hs-memory
+    tag: 296b79424854eae293f6ba09b5308a0bf4dfd6d5
```
Comment (parsonsmatt) [inline cabal.project:59]: This can be dropped now: https://hackage.haskell.org/package/memory-0.18.0/changelog

**Comment by TeofilC** [comment]
Just a quick update on this. We are still waiting on sol/doctest to support 9.4 as well as the postgresql packages to bump bounds.

**Comment by TeofilC** [comment]
I've rebased on top of master and updated the cabal.project to use sol/doctest HEAD, which supports 9.4. I have the doctests passing locally now. We are waiting on some hackage releases/ version bumps in order to be able to drop stuff from the cabal.project.

It looks like there is a genuine failure on the ghcjs 8.6 test suite because it doesn't seem to support the type applications syntax that I'm using to fix stuff for 9.4. There's several ways we could fix this:
1) drop support for ghcjs 8.6
2) hide the type applications behind CPP, so, it only occurs for GHC 9.4+
3) try to figure out some way of avoiding this issue without using type applications

**Comment by tchoutri** [comment]
After some thought, I vote in favour of 2.

**Comment by TeofilC** [comment]
I've added a commit that implements that option.

File: cabal.project [hunk 4]
```diff
@@ -56,9 +56,4 @@ allow-newer:
 -- see https://github.com/haskellari/postgresql-simple/issues/104
     postgresql-simple:base
   , postgresql-simple:template-haskell
--- https://github.com/haskell-servant/servant-js/pull/49
-  , servant-js:base
-  , servant-js:lens
-  , servant-js:text
--- https://github.com/haskell-servant/servant-multipart/pull/64
-  , servant-multipart:lens
+  , swagger2:base
```
Comment (maksbotan) [inline cabal.project:59]: did you submit a PR there?

File: .github/workflows/master.yml [hunk 5]
```diff
@@ -20,6 +20,7 @@ jobs:
           - "8.10.7"
           - "9.0.2"
           - "9.2.2"
+          - "9.4.1"
```
[the remark this answers is not in the document]
Comment (TeofilC) [reply .github/workflows/master.yml]: Done in b49a2b2ed4a4e0caa18f71909d0024bac76c00eb

File: cabal.project (hunk 4 above)
Comment (TeofilC) [reply cabal.project:60]: I've made a PR now: https://github.com/GetShopTV/swagger2/pull/244
Link added in1c2299be62abc0c48997f9bed0157635dfb50202

**Comment by TeofilC** [comment]
I'll take a look at the doctest failures tomorrow

File: cabal.project (hunk 4 above)
Comment (maksbotan) [reply cabal.project:60]: It got released: https://hackage.haskell.org/package/swagger2-2.8.6

**Review by maksbotan — COMMENTED** [review]
Looks like CI for 9.4 failed

File: cabal.project (hunk 4 above)
Comment (TeofilC) [reply cabal.project]: Override removed in 91269730729a39d757d111e311b0490029b16b40

**Comment by TeofilC** [comment]
The doctest failures are caused by GHC's error output changing.

It used to produce message that start with `Could not deduce` but now they start with `Could not solve`.

I'm going to update the tests to just expect `Could not` as that's common to both and still ensures that some sort of error's been triggered. See 86c61c6dbdc82ffcab5803e15411f7fce82874b1

**Comment by maksbotan** [comment]
Thanks!

Published Hackage revisions:
- servant-client-core
- servant-client
- servant-auth-client
- servant-auth-docs
- servant-auth-server
- servant-auth-swagger
- servant-docs
- servant-foreign
- servant-http-streams

New versions:
- servant-server 0.19.2
- servant 0.19.1

**Comment by benz0li** [comment]
Thank you @TeofilC

**Pull request merged**